java/arcs/core/storage/CapabilitiesResolverNew.kt

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage
+
+import arcs.core.common.ArcId
+import arcs.core.data.CapabilitiesNew
+import arcs.core.data.CollectionType
+import arcs.core.data.EntityType
+import arcs.core.data.ReferenceType
+import arcs.core.data.Schema
+import arcs.core.data.SingletonType
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+import arcs.core.type.Type
+
+/**
+ * [CapabilitiesResolverNew] is a factory class that creates [StorageKey]s based on the registered
+ * providers and given [CapabilitiesNew].
+ */
+class CapabilitiesResolverNew(
+    val options: Options,
+    factoriesList: List<StorageKeyFactory> = listOf(),
+    val selector: FactorySelector = SimpleCapabilitiesSelector()
+) {
+    val factories: Map<String, StorageKeyFactory>
+
+    init {
+        require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
+            "Storage keys protocol must be unique $factoriesList."
+        }
+        factories = factoriesList.map { it.protocol to it }.toMap().toMutableMap()
+        CapabilitiesResolverNew.defaultStorageKeyFactories.forEach { (protocol, factory) ->
+            if (factories[protocol] == null) factories.put(protocol, factory)
+        }
+    }
+
+    /* Options used to construct [CapabilitiesResolver] */
+    data class Options(val arcId: ArcId)
+
+    fun createStorageKey(capabilities: CapabilitiesNew, type: Type, handleId: String): StorageKey {
+        val selectedFactories =
+            // defaultStorageKeyFactories.filterValues { it.supports(capabilities) } +
+            factories.filterValues { it.supports(capabilities) }
+
+        if (selectedFactories.isEmpty()) {
+            throw IllegalStateException("Cannot create storage key for handle '$handleId' with " +
+                "capabilities $capabilities")
+        }
+        val factory = selector.select(selectedFactories.values)
+        return createStorageKeyWithFactory(requireNotNull(factory), type, handleId)
+    }
+
+    fun createStorageKeyWithFactory(
+        factory: StorageKeyFactory,
+        type: Type,
+        handleId: String
+    ): StorageKey {
+        val containerKey = factory.create(StorageKeyFactory.ContainerStorageKeyOptions(
+            options.arcId, toEntitySchema(type)))
```

Comment:
Nit: line wrapping:

```
val containerKey = factory.create(
    StorageKeyFactory.ContainerStorageKeyOptions(options.arcId, toEntitySchema(type))
)
```

Or if that still doesn't fit:

```
val containerKey = factory.create(
    StorageKeyFactory.ContainerStorageKeyOptions(
        options.arcId,
        toEntitySchema(type)
    )
)
```

Reply:
done